river/tree/streaming_gradient_tree.py

```diff
@@ -0,0 +1,460 @@
+
+    Parameters
+    ----------
+    delta
+        Define the significance level of the F-tests performed to decide upon creating splits
+        or updating predictions.
+    grace_period
+        Interval between split attempts or prediction updates.
+    init_pred
+        Initial value predicted by the tree.
+    max_depth
+        The maximum depth the tree might reach. If set to `None`, the trees will grow
+        indefinitely.
+    lambda_value
+        Positive float value used to impose a penalty over the tree's predictions and force
+        them to become smaller. The greater the lambda value, the more constrained are the
+        predictions.
+    gamma
+        Positive float value used to impose a penalty over the tree's splits and force them to
+        be avoided when possible. The greater the gamma value, the smaller the chance of a
+        split occurring.
+    nominal_attributes
+        List with identifiers of the nominal attributes. If None, all features containing
+        numbers are assumed to be numeric.
+    quantization_strategy
+        Defines the policy used to define the quantization radius applied on numerical
+        attributes' discretization. Each time a new leaf is created, a radius, or
+        discretization interval, is assigned to each feature in order to quantize it.
+        The smaller the interval used, the more points are going to be stored by the tree
+        between splits. Although using more points ought to increase the memory footprint and
+        runtime, the split candidates potentially are going to be better. When the trees are
+        created and no data was observed, the initial radius is defined by the parameter
+        `default_radius`. The next nodes are going to use the policy define by this
+        parameter.</br>
+        Valid values are:
+        - 'stddiv': use the standard deviation of the feature divided by `quantization_radius_div`
+        as the radius.</br>
+        - 'constant': use the value defined in 'default_radius' as the static quantization
+        interval.
+    quantization_radius_div
+        Value by which the standard deviation of numeric features is going to be divided to define
+        the quantization radius. Only used when `quantization_strategy='stddiv'`.
+    default_radius
+        Quantization radius used when the tree is created and no data is observed or when
+        `quantization_strategy='constant'`.
+
+    Examples
+    --------
+    >>> from river import datasets
+    >>> from river import evaluate
+    >>> from river import metrics
+    >>> from river import tree
+
+    >>> dataset = datasets.Phishing()
+    >>> model = tree.StreamingGradientTreeClassifier()
+    >>> metric = metrics.Accuracy()
+
+    >>> evaluate.progressive_val_score(dataset, model, metric)
+    Accuracy: 82.32%
+
+    Notes
+    -----
+    This implementation enhances the original proposal [^1] by using an incremental strategy to
```

Comment:
This needs to be confirmed by Henry. Can we make this optional, and use as default the "vanilla" version?

Reply:
It makes sense. @henrygouk, how do you envision the calls to `predict_one` happening during the warm-up period? Should we always return the initial prediction during this time? For instance, always return zero in the regression case and `False` in the classification case?

One of my motivations to modify the SGTs with my proposal (QO) was to avoid this warm-up period from delaying learning and degrading the prediction performance. 

Any feedback is welcomed!